### PR TITLE
default meth=0 (ES) in spread/interp test drivers; new make spreadtest task

### PR DESCRIPTION
**Pull request by ahbarnett**
This
* sets meth=0 (instead of 1) as default in the spread/interp standalone test drivers.
* adds spreadperf.sh bash script to quickly time spread/interp drivers at density=1 medium-sized problem.
* in Makefile splits out the build of spread/interp testers from the make all task, adds own task.

Note that this exposes a bug in interp3d_test. Some cudafree thing, probably easy to fix.
See output of `make spreadtest`:

```
(base) ahb@workergpu23 /mnt/home/ahb/numerics/cufinufft> make spreadtest
Running spread/interp only tests...
(cd test; ./spreadperf.sh)
spread 2D..............................................
[Method 1] 1000000 NU pts to #1000000 U pts in 0.0242 s (4.13e+07 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.0088 s (1.14e+08 NU pts/s)
[Method 1] 1000000 NU pts to #1000000 U pts in 0.00438 s (2.28e+08 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.00275 s (3.63e+08 NU pts/s)
interp 2D..............................................
[Method 1] 1000000 U pts to #1000000 NU pts in 0.00323 s (	3.09e+08 NU pts/s)
[Method 2] 1000000 U pts to #1000000 NU pts in 0.00387 s (	2.58e+08 NU pts/s)
[Method 1] 1000000 U pts to #1000000 NU pts in 0.00186 s (	5.37e+08 NU pts/s)
[Method 2] 1000000 U pts to #1000000 NU pts in 0.00233 s (	4.3e+08 NU pts/s)
spread 3D..............................................
[Method 1] 1000000 NU pts to #1000000 U pts in 0.518 s (1.93e+06 NU pts/s)
[Method 1] 1000000 NU pts to #1000000 U pts in 0.0334 s (3e+07 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.0107 s (9.31e+07 NU pts/s)
interp 3D..............................................
CUDA error at src/memtransfer_wrapper.cu:353 code=77(cudaErrorIllegalAddress) "cudaFree(d_plan->idxnupts)" 
CUDA error at src/memtransfer_wrapper.cu:353 code=77(cudaErrorIllegalAddress) "cudaFree(d_plan->idxnupts)" 
CUDA error at src/memtransfer_wrapper.cu:364 code=77(cudaErrorIllegalAddress) "cudaFree(d_plan->idxnupts)" 
make: *** [spreadtest] Error 1
````

### CODE REVIEW

**Comment by ahbarnett** [comment]
sorry, forgot to pull on the machine I edited on. Fixing now.

**Comment by ahbarnett** [comment]
ok, now I merged w/ latest master (silly me earlier). Now onyl spread2d works; others have free sortidx problems.
Maybe Melody you have insight:

```
(base) ahb@workergpu23 /mnt/home/ahb/numerics/cufinufft> make spreadtest
Running spread/interp only tests...
(cd test; ./spreadperf.sh)
spread 2D..............................................
[Method 1] 1000000 NU pts to #1000000 U pts in 0.0242 s (4.13e+07 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.00872 s (1.15e+08 NU pts/s)
[Method 1] 1000000 NU pts to #1000000 U pts in 0.00342 s (2.92e+08 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.00233 s (4.28e+08 NU pts/s)
interp 2D..............................................
CUDA error at src/memtransfer_wrapper.cu:113 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:113 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:113 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:113 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
spread 3D..............................................
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
interp 3D..............................................
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
CUDA error at src/memtransfer_wrapper.cu:334 code=17(cudaErrorInvalidDevicePointer) "cudaFree(d_plan->sortidx)" 
make: *** [spreadtest] Error 1
```

**Comment by ahbarnett** [comment]
The clues are
1) that spread2d_test works, and the other three don't
(unless M has a certain range, eg for interp2d_test the default M does work....).

2) that spread2d_test uses sort_gpu =0 (which seems bad - we want to test
sorting on the gpu not cpu). But I changed to 1, and it still works.

I can't find where `d_plan->sortidx` is allocated.  Ah, got it...
The problem is that there are tests whether these d_plan objects have been allocated or not; however, they need to be initialized as NULL in all plan/testers.
I'll do that.

**Comment by ahbarnett** [comment]
ok, made sure spread/interp testers alloc dplan to NULL, otherwise the dealloc check in plan may try to dealloc a random uninitialized ptr value (was the cause of cuda errors).
Here's the correct result for `make spreadtest` :

```
Running spread/interp only tests...
(cd test; ./spreadperf.sh)
spread 2D..............................................
[Method 1] 1000000 NU pts to #1000000 U pts in 0.00987 s (1.01e+08 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.0087 s (1.15e+08 NU pts/s)
[Method 1] 1000000 NU pts to #1000000 U pts in 0.00332 s (3.01e+08 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.00233 s (4.3e+08 NU pts/s)
interp 2D..............................................
[Method 1] 1000000 U pts to #1000000 NU pts in 0.00338 s (	2.96e+08 NU pts/s)
[Method 2] 1000000 U pts to #1000000 NU pts in 0.00377 s (	2.66e+08 NU pts/s)
[Method 1] 1000000 U pts to #1000000 NU pts in 0.00187 s (	5.35e+08 NU pts/s)
[Method 2] 1000000 U pts to #1000000 NU pts in 0.00203 s (	4.93e+08 NU pts/s)
spread 3D..............................................
[Method 1] 1000000 NU pts to #1000000 U pts in 0.115 s (8.71e+06 NU pts/s)
[Method 1] 1000000 NU pts to #1000000 U pts in 0.0118 s (8.46e+07 NU pts/s)
[Method 2] 1000000 NU pts to #1000000 U pts in 0.0077 s (1.3e+08 NU pts/s)
interp 3D..............................................
[Method 1] 1000000 U pts to #1000000 NU pts in 0.0509 s (	1.97e+07 NU pts/s)
[Method 1] 1000000 U pts to #1000000 NU pts in 0.00605 s (	1.65e+08 NU pts/s)
[Method 2] 1000000 U pts to #1000000 NU pts in 0.00522 s (	1.91e+08 NU pts/s)
```

**Comment by ahbarnett** [comment]
Jenkins was failing since I hadn't realised you had no default make task set (in which case the order of the tasks matters in makefile!). Appears CI just runs `make` so the default task is crucial. I am confused why it doesn't run `make check`.
Anyway, since this PR changes the makefile a bit (in particular, adding spreadtest to the task `all', which is now run as part of CI), I'm requesting y'all as reviewers and someone can pull it in.

Jenkins fails with
```
libcuda.so.1: cannot open shared object file: No such file or directory
```
errors when executing the new tests I had in the makefile.   Maybe we had not got to the stage of running executables in CI?
In which case, tell me to remove `spreadtest` from `all`, for now.

File: Makefile [hunk 1]
```diff
@@ -122,17 +120,24 @@ all: $(BINDIR)/spread2d_test \
...
+spreadtest: $(BINDIR)/spread2d_test \
+	$(BINDIR)/spread2d_test_32 \
+	$(BINDIR)/interp2d_test \
+	$(BINDIR)/interp2d_test_32 \
+	$(BINDIR)/spread3d_test \
+	$(BINDIR)/spread3d_test_32 \
+	$(BINDIR)/interp3d_test \
+	$(BINDIR)/interp3d_test_32
+	@echo "Running spread/interp only tests..."
+	(cd test; ./spreadperf.sh)
```
Comment (MelodyShih) [inline Makefile:140]: There were some spread/interp only tests in `check` section. With this change, I think we can remove them there.

**Review by MelodyShih — CHANGES_REQUESTED** [review]
Apart from the failing of Jenkins test and the comment on the spread/interp test in the `check` section, the changes look good to me. 
Thanks for catching the noninitialized ptr value error I made.

File: Makefile (hunk 1 above)
Comment (ahbarnett) [reply Makefile]: I left these in, but rationalized all targets in a better way.

**Comment by ahbarnett** [comment]
Dear all, I cleaned up the Makefile. There is now clear separation between compile vs execution (checks). @garrettwrong should check this, and also update CI so that it executes something (!). Right now it only compiles.  However, when I tried to add execution CI failed with cuda .so's not found.

**Review by MelodyShih — APPROVED** [review]

**Review by garrettwrong — APPROVED** [review]
I checked this out locally.  `make` and `make check` appear to work.  This is better than it was.

I can look at the CI stuff, but maybe should be separate issues.  There are two problems. First, centos6 is deprecated so the centos7 recipe may need to be completed instead, or at least the centos6 updated for posterity.  Second, the CI runs the python tests, which are not comprehensive, but do test basic function... running make check is possible, but it should probably be an additional stage, and before python. I just wanted to point out that some smoke tests are run against the build, just wrapped up in python.... 

Neither of those need to hold up this PR.

**Pull request merged**